function/graphite/graphite_test.go

```diff
@@ -0,0 +1,300 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package graphite
+
+import (
+	"testing"
+
+	"github.com/square/metrics/api"
+	"github.com/square/metrics/internal"
+)
+
+func TestApplyPattern(t *testing.T) {
+	tests := []struct {
+		pattern string
+		metric  string
+		success bool
+		expect  api.TagSet
+	}{
+		{
+			pattern: "this.is.a.graphite.metric",
+			metric:  "this.is.a.graphite.metric",
+			success: true,
+			expect:  api.TagSet{"$graphite": "this.is.a.graphite.metric"},
+		},
+		{
+			pattern: "this.is.a.graphite.metric",
+			metric:  "this.is.a.different_graphite.metric",
+			success: false,
+		},
+		{
+			pattern: "this.is.a.graphite.metric",
+			metric:  "this.is.a.different.graphite.metric",
+			success: false,
+		},
+		{
+			pattern: "this.is.a.graphite.metric",
+			metric:  "this.is.a.graphite.metric.too",
+			success: false,
+		},
+		{
+			pattern: "this.is.a.graphite.metric",
+			metric:  "this.is.a.graphite.metricQ",
+			success: false,
+		},
+		{
+			pattern: "this.is.a.%something%.metric",
+			metric:  "this.is.a.graphite.metric",
+			success: true,
+			expect: api.TagSet{
+				"$graphite": "this.is.a.graphite.metric",
+				"something": "graphite",
+			},
+		},
+		{
+			pattern: "this.is.a.%something%.%type%",
+			metric:  "this.is.a.graphite.metric",
+			success: true,
+			expect: api.TagSet{
+				"$graphite": "this.is.a.graphite.metric",
+				"something": "graphite",
+				"type":      "metric",
+			},
+		},
+		{
+			pattern: "%word1%.%word2%.%word3%.%word4%.%word5%",
+			metric:  "this.is.a.graphite.metric",
+			success: true,
+			expect: api.TagSet{
+				"$graphite": "this.is.a.graphite.metric",
+				"word1":     "this",
+				"word2":     "is",
+				"word3":     "a",
+				"word4":     "graphite",
+				"word5":     "metric",
+			},
+		},
+		{
+			pattern: "%app%.%datacenter%.cpu.%quantity%",
+			metric:  "metrics-query-engine.north.cpu.total",
+			success: true,
+			expect: api.TagSet{
+				"$graphite":  "metrics-query-engine.north.cpu.total",
+				"app":        "metrics-query-engine",
+				"datacenter": "north",
+				"quantity":   "total",
+			},
+		},
+	}
+	for i, test := range tests {
+		rule, err := internal.Compile(internal.RawRule{Pattern: test.pattern, MetricKeyPattern: "$graphite"})
+		if (err != nil) && test.success {
+			t.Errorf("Expected success, but rule compilation failed for test %+v", test)
+		}
+		if err != nil {
+			continue
+		}
+		result, ok := applyPattern(rule, test.metric)
+		if ok != test.success {
+			t.Errorf("Test i = %d, test = %+v: didn't expect ok = %t", i, test, ok)
+			continue
+		}
+		if !test.expect.Equals(result.TagSet) {
+			t.Errorf("Expected %+v but got %+v", test.expect, result)
+		}
+	}
+}
+
+type testStore struct {
```

Comment:
if you just embed the interface into it, you wouldn't need to implement (most) of the methods below.